src/backend/distributed/planner/multi_logical_optimizer.c

```diff
@@ -1821,41 +1879,143 @@ AddTypeConversion(Node *originalAggregate, Node *newExpression)
 
 /*
  * WorkerExtendedOpNode creates the worker extended operator node from the given
- * target entries. The function walks over these target entries; and for entries
- * with aggregates in them, this function calls the recursive aggregate walker
- * function to create aggregates for the worker nodes. Also, the function checks
- * if we can push down the limit to worker nodes; and if we can, sets the limit
- * count and sort clause list fields in the new operator node. It provides special
- * treatment for count distinct operator if it is used in repartition subqueries
- * or on non-partition columns. Each column in count distinct aggregate is added
- * to target list, and group by list of worker extended operator.
+ * originalOpNode and extendedOpNodeProperties.
+ *
+ * For the details of the processing see the comments of the functions that
+ * are called from this function.
  */
 static MultiExtendedOp *
 WorkerExtendedOpNode(MultiExtendedOp *originalOpNode,
 					 ExtendedOpNodeProperties *extendedOpNodeProperties)
 {
-	MultiExtendedOp *workerExtendedOpNode = NULL;
-	List *targetEntryList = originalOpNode->targetList;
-	ListCell *targetEntryCell = NULL;
-	List *newTargetEntryList = NIL;
-	List *groupClauseList = copyObject(originalOpNode->groupClauseList);
-	Node *havingQual = originalOpNode->havingQual;
+	MultiExtendedOp *workerExtendedOpNode = CitusMakeNode(MultiExtendedOp);
 	AttrNumber targetProjectionNumber = 1;
-	WorkerAggregateWalkerContext *walkerContext =
-		palloc0(sizeof(WorkerAggregateWalkerContext));
 	Index nextSortGroupRefIndex = 0;
-	bool queryHasAggregates = false;
-	bool distinctClauseSupersetofGroupClause = false;
 	bool distinctPreventsLimitPushdown = false;
-	bool createNewGroupByClause = false;
+	bool groupByExtended = false;
 	bool groupedByDisjointPartitionColumn =
+
 		extendedOpNodeProperties->groupedByDisjointPartitionColumn;
-	bool pushDownWindowFunction = extendedOpNodeProperties->pushDownWindowFunctions;
 
-	walkerContext->expressionList = NIL;
-	walkerContext->pullDistinctColumns = extendedOpNodeProperties->pullDistinctColumns;
+	List *originalTargetEntryList = originalOpNode->targetList;
+	List *originalGroupClauseList = originalOpNode->groupClauseList;
+	List *originalSortClauseList = originalOpNode->sortClauseList;
+	Node *originalHavingQual = originalOpNode->havingQual;
+	Node *originalLimitCount = originalOpNode->limitCount;
+	Node *originalLimitOffset = originalOpNode->limitOffset;
+	List *originalWindowClause = originalOpNode->windowClause;
+	List *originalDistinctClause = originalOpNode->distinctClause;
+	bool hasDistinctOn = originalOpNode->hasDistinctOn;
+
+	int originalGroupClauseLength = list_length(originalGroupClauseList);
+	bool queryHasAggregates = TargetListHasAggragates(originalTargetEntryList);
+
+	/* calculate the next sort group index based on the original target list */
+	nextSortGroupRefIndex = GetNextSortGroupRef(originalTargetEntryList);
```

Comment:
The use of `targetProjectNumber` and `nextSortGroupRefIndex` keeps bugging me a bit. Even though they are essentially simple counters of how much stuff we added to the target list and group by/order by, it's hard to figure out what they represent from their name, or the code, or the comments. It's also a bit hard to figure out which variables belong together and what their relationships are.

I think the quickest way to resolve that is to add a comment saying what the variable represents to help the reader digest the parameters. A more structural approach could be to just move these variables into the `MultiExtendedOpNode` and passing the full `workerExtendedOpNode` in the functions below. Even though they are not useful in the physical planner, the improved simplicity/readability would be worth having some redundant fields.

Reply:
I'll be unifying multiple comments from you:

>  It's also a bit hard to figure out which variables belong together and what their relationships are.
and
>Given the number of workerExtendedOpNode fields we pass, I'm wondering whether we shouldn't just pass workerExtendedOpNode

Thinking about your comments, I'd propose the following:

* We create few new smaller structs
* We'd be able to decrease the number of parameters passed to many functions
* We'd be able to create a relationship between the two indexes and their corresponding data structures (note that `nextSortGroupRefIndex` might be in multiple places)

```typdef struct QueryTargetList
{
List **newTargetEntryList;
AttrNumber *targetProjectionNumber;
}

typdef struct QueryGroupClause
{
List **groupClauseList;
AttrNumber *nextSortGroupRefIndex;
}

typedef struct QuerySortLimitClauses
{
Node **workerLimitCount;
List **workerSortClauseList
AttrNumber *nextSortGroupRefIndex;
}

```

@marcocitus  Since the above is a relatively large change in the PR, I'll wait confirmation from you before doing any changes.

Reply:
I think this makes a lot of sense and is way better than my proposal.

Reply:
I pushed the changes. I think now it looks (much) better. What do you think?

`OrderByLimit` processing very tightly coupled, so that part still looks a bit crowded. Still, that seems acceptable for now.